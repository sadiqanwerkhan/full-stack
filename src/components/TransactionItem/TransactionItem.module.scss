@import '../../styles/variables.scss';

.grid-li {
    display: grid;
    grid-template-columns: 36px min-content 1fr min-content 24px;
    grid-gap: $spacing-mega;
    justify-items: start;
}