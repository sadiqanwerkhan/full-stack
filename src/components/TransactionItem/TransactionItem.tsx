import React, { FC } from "react";

import { Currency } from "../Currency";
import { Badge } from "../Badge";
import { formatDate } from "../../services/DateService";
import {
  PaymentMethods,
  PaymentStatus,
} from "../../constants/TransactionConstants";

import styles from "./TransactionItem.module.scss";

const PAYMENT_METHOD_ICONS = {
  [PaymentMethods.CASH]: "ðŸ’µ",
  [PaymentMethods.CARD]: "ðŸ’³",
};

export type Transaction = {
  id: number;
  date: Date;
  amount: number;
  status: PaymentStatus;
  currency: string;
  paymentMethod: PaymentMethods;
  paymentType: "card" | "cash";
};

interface TransactionItemProps {
  transaction: Transaction;
}

const TransactionItem: FC<TransactionItemProps> = ({ transaction }) => {
  return (
    <li className={styles["grid-li"]}>
      <div>#{transaction.id}</div>
      <div>{formatDate(transaction.date)}</div>
      <Badge status={transaction.status}>{transaction.status}</Badge>
      <Currency
        paymentStatus={transaction.status}
        currencyCode={transaction.currency}
        amount={transaction.amount}
      />
      <span aria-label={`Paid by ${transaction.paymentMethod}`}>
        {PAYMENT_METHOD_ICONS[transaction.paymentMethod]}
      </span>
    </li>
  );
};

export { TransactionItem };
