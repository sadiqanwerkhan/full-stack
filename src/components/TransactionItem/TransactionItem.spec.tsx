import "@testing-library/jest-dom";

import React from "react";
import { render } from "@testing-library/react";

import { TransactionItem, Transaction } from "./TransactionItem";
import {
  PaymentMethods,
  PaymentStatus,
} from "@/constants/TransactionConstants";

const transaction: Transaction = {
  id: 1,
  date: new Date("2020-01-11T18:32:32.605Z"),
  amount: 14425.17,
  status: PaymentStatus.Refunded,
  currency: "EUR",
  paymentMethod: PaymentMethods.CARD,
};

describe("Transaction Items", () => {
  describe("with card payment", () => {
    it("should match styles", () => {
      const { asFragment } = render(
        <TransactionItem transaction={transaction} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("with cash payment", () => {
    it("should match styles", () => {
      const { asFragment } = render(
        <TransactionItem
          transaction={{ ...transaction, paymentMethod: PaymentMethods.CASH }}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
