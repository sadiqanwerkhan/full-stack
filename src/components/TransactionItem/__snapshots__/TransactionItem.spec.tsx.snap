// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transaction Items with card payment should match styles 1`] = `
<DocumentFragment>
  <li
    class="grid-li"
  >
    <div>
      #1
    </div>
    <div>
      11/01/2020
    </div>
    <div
      class="badge badge-refunded"
    >
      refunded
    </div>
    <span>
      -14.425,17Â â‚¬
    </span>
    <span
      aria-label="Paid by card"
    >
      ðŸ’³
    </span>
  </li>
</DocumentFragment>
`;

exports[`Transaction Items with cash payment should match styles 1`] = `
<DocumentFragment>
  <li
    class="grid-li"
  >
    <div>
      #1
    </div>
    <div>
      11/01/2020
    </div>
    <div
      class="badge badge-refunded"
    >
      refunded
    </div>
    <span>
      -14.425,17Â â‚¬
    </span>
    <span
      aria-label="Paid by cash"
    >
      ðŸ’µ
    </span>
  </li>
</DocumentFragment>
`;
